# Module 4 Lab 2: Textbook Problem 3.1

## The Bookstore Inventory

This lab uses the bookstore inventory data to find the median cover type, the mean weight of all books, and the mean weight for each cover type.

**GitHub repository:** https://github.com/roee0910/cs82a-portfolio

The data file contains 15 books. The `weight` column is measured in grams, and `cover` identifies the book as hardback (`hb`) or paperback (`pb`).

In [1]:
import pandas as pd

books = pd.read_csv("allbacks.csv")
books.head()

,rownames,volume,area,weight,cover
0,1,885,382,800,hb
1,2,1016,468,950,hb
2,3,1125,387,1050,hb
3,4,239,371,350,hb
4,5,701,371,750,hb


## 1. Median cover type

Because the cover types are ordered categories (`hb` comes before `pb`), I can sort them and take the middle value. There are 15 books, so the median is the 8th value after sorting.

In [2]:
cover_sorted = books["cover"].sort_values().reset_index(drop=True)

print("Sorted cover types:")
print(cover_sorted.to_string(index=False))
print("\nNumber of books:", len(cover_sorted))
print("Middle position:", (len(cover_sorted) + 1) // 2)
print("Median cover type:", cover_sorted.iloc[(len(cover_sorted) - 1) // 2])

Sorted cover types:
hb
hb
hb
hb
hb
hb
hb
pb
pb
pb
pb
pb
pb
pb
pb

Number of books: 15
Middle position: 8
Median cover type: pb


The median cover type is **pb (paperback)** because the 8th cover type in the ordered list is `pb`.

In [3]:
median_cover = cover_sorted.iloc[(len(cover_sorted) - 1) // 2]
median_cover

'pb'

## 2. Mean weight of all books

The overall mean weight is the total weight of all 15 books divided by 15.

In [4]:
overall_mean_weight = books["weight"].mean()

print("Total weight:", books["weight"].sum(), "grams")
print("Number of books:", books["weight"].count())
print("Mean weight:", overall_mean_weight, "grams")

Total weight: 10600 grams
Number of books: 15
Mean weight: 706.6666666666666 grams


The mean weight of all the books is **706.67 grams**. This means a typical book in this inventory weighs about 707 grams on average.

In [5]:
overall_mean_weight

np.float64(706.6666666666666)

## 3. Mean weight by cover type

Next, I grouped the books by cover type and calculated the mean weight for each group.

In [6]:
cover_means = books.groupby("cover")["weight"].mean()

print(cover_means.round(2))

cover
hb    796.43
pb    628.12
Name: weight, dtype: float64


The mean weight is **796.43 grams for hardback books** and **628.13 grams for paperback books**. Hardback books are about **168.30 grams heavier on average** than paperback books, which suggests that the type of cover is associated with a noticeable difference in book weight in this inventory.

In [7]:
difference = cover_means["hb"] - cover_means["pb"]

print("Hardback mean:", round(cover_means["hb"], 2), "grams")
print("Paperback mean:", round(cover_means["pb"], 2), "grams")
print("Difference:", round(difference, 2), "grams")

Hardback mean: 796.43 grams
Paperback mean: 628.12 grams
Difference: 168.3 grams


## 4. By-hand verification

I verified the overall mean using the formula:

**Mean = total of all weights ÷ number of books**

The total weight is 10,600 grams and there are 15 books, so:

**10,600 ÷ 15 = 706.67 grams**

This matches the Python result of 706.67 grams.

For the median, there are 15 ordered cover values, so the middle position is the 8th value. The first 7 are `hb` and the 8th is `pb`, so the median cover type is `pb`, which also matches the Python result.

## Final answers

- **Median cover type:** `pb` (paperback)
- **Mean weight of all books:** **706.67 grams**
- **Mean weight of hardback books:** **796.43 grams**
- **Mean weight of paperback books:** **628.13 grams**
- **Difference between group means:** **168.30 grams**, with hardback books being heavier on average.

The calculations and the by-hand checks agree, and the group comparison suggests that hardback books in this inventory tend to weigh more than paperback books.